go/chat/journey_card_manager.go

```diff
@@ -9,43 +9,139 @@ import (
 
 	lru "github.com/hashicorp/golang-lru"
 	"github.com/keybase/client/go/chat/globals"
+	"github.com/keybase/client/go/chat/storage"
 	"github.com/keybase/client/go/chat/types"
 	"github.com/keybase/client/go/chat/utils"
+	"github.com/keybase/client/go/encrypteddb"
 	"github.com/keybase/client/go/libkb"
 	"github.com/keybase/client/go/protocol/chat1"
 	"github.com/keybase/client/go/protocol/gregor1"
 	"github.com/keybase/client/go/protocol/keybase1"
 )
 
-type JourneyCardChecker struct {
+// JourneyCardManager handles user switching and proxies to the active JourneyCardManagerSingleUser.
+type JourneyCardManager struct {
 	globals.Contextified
 	utils.DebugLabeler
-	storageLock sync.Mutex
-	lru         *lru.Cache // TODO in-memory storage is insufficient, add leveldb storage.
-	// TODO when it comes to disk storage. Do versioning in a way that makes it easy to reset. But not spam cards. Maybe burn card types for a conv.
+	switchLock sync.Mutex
+	m          *JourneyCardManagerSingleUser
 }
 
-var _ (types.JourneyCardManager) = (*JourneyCardChecker)(nil)
+var _ (types.JourneyCardManager) = (*JourneyCardManager)(nil)
+
+func NewJourneyCardManager(g *globals.Context) *JourneyCardManager {
+	return &JourneyCardManager{
+		Contextified: globals.NewContextified(g),
+		DebugLabeler: utils.NewDebugLabeler(g.GetLog(), "JourneyCardManager", false),
+	}
+}
+
+func (j *JourneyCardManager) get(ctx context.Context, uid gregor1.UID) (*JourneyCardManagerSingleUser, error) {
+	if uid.IsNil() {
+		return nil, fmt.Errorf("missing uid")
+	}
+	err := libkb.AcquireWithContextAndTimeout(ctx, &j.switchLock, 10*time.Second)
+	if err != nil {
+		return nil, fmt.Errorf("JourneyCardManager switchLock error: %v", err)
+	}
+	defer j.switchLock.Unlock()
+	if j.m != nil && !j.m.uid.Eq(uid) {
+		j.m = nil
```

Comment:
is this all that is required to clean up a journeycardmanagersingleuser?

Reply:
It doesn't have any background processes and its leveldb storage is keyed by uid. I thought so, but no. `storage.GetSecretBoxKey` gets the db key for whoever is currently logged in. I'll see if there's a way to pin that.

Reply:
now secret key must be for the known `keyFn` so should be good